### PR TITLE
Bug 1992596: e2e/cli: move half of builds.sh to proper e2e

### PR DESCRIPTION
**Pull request by atiratree**
I found there is a bit of overlap with `[sig-builds]` as they are using oc cli to create builds as well. But it does not test the CLI part thourougly. I am not entirerly sure if it makes sense to combine `sig-builds` with `sig-cli`, but I decided not to at this moment.

### CODE REVIEW

File: test/extended/cli/builds.go
Comment (atiratree) [reply test/extended/cli/builds.go]: I went with the function for now - we can change it later

File: test/extended/cli/builds.go
Comment (atiratree) [reply test/extended/cli/builds.go]: resolved

File: test/extended/cli/builds.go
Comment (ravisantoshgudimetla) [inline test/extended/cli/builds.go:77]: I'd rather have a `g.It` instead of simple `g.By` to explicitly call out the functionality we are testing.

File: test/extended/cli/builds.go
Comment (ravisantoshgudimetla) [inline test/extended/cli/builds.go:32]: We can move this function to `g.JustBeforeEach` as a prerequisite step when each `g.By` becomes a `g.It`

File: test/extended/cli/builds.go
Comment (ravisantoshgudimetla) [inline test/extended/cli/builds.go:59]: The cleanup can be moved to `g.AfterEach()`

File: test/extended/cli/builds.go
Comment (ravisantoshgudimetla) [inline test/extended/cli/builds.go:47]: What would happen if opts.bcName is empty? Would we error out or ignore?

I'd rather be explicit in the individual test rather than trying to get arg and then do validation here.

File: test/extended/cli/builds.go
Comment (ravisantoshgudimetla) [inline test/extended/cli/builds.go:36]: Shouldn't we wait for the build? Isn't the whole purpose of moving to go code is to be reasonable with the time frame and have knobs to control it?

**Review by ravisantoshgudimetla — CHANGES_REQUESTED** [review]
Thank you for working on this @atiratree. The code looks correct to me. However I have the following concerns 
- The tests are not explicit. It'd also be hard to parse the [output](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/26141/pull-ci-openshift-origin-master-e2e-gcp/1391681995452452864/artifacts/e2e-gcp/openshift-e2e-test/artifacts/e2e.log) when everything passes 
- The generic method `testNewBuild` is trying to be too generic. I believe you've gone with this approach since you did not want to repeat the code block. However in case of tests, I'd rather be explicit and be repetitive if it helps in debugging.
- I think we should wait for the build to finish and throw an error

**Comment by atiratree** [comment]
> Thank you for working on this @atiratree. The code looks correct to me. However I have the following concerns
> 
> * The tests are not explicit. It'd also be hard to parse the [output](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/26141/pull-ci-openshift-origin-master-e2e-gcp/1391681995452452864/artifacts/e2e-gcp/openshift-e2e-test/artifacts/e2e.log) when everything passes
> * The generic method `testNewBuild` is trying to be too generic. I believe you've gone with this approach since you did not want to repeat the code block. However in case of tests, I'd rather be explicit and be repetitive if it helps in debugging.

I wasn't sure with too many `It`s, since it generate a lot of new tests. 

Originally, I thought it would not be so bad for debugging - since one can still infer the place where bug happened from the logs. Although I agree it is a bit harder.

So, it is generally not advisible to do logic extraction? 
Would `for` which generates multiple `It`s help? Or can I extract some of the checks into functions? Or should all the logic be copy pasted? 

> * I think we should wait for the build to finish and throw an error

That sounds like a good idea. I wasn't sure how deep I should go. If I should just rewrite the tests. Or if I should analyze the CLI functionality regarding builds completely and enhance the tests? It would take considerably longer though.

File: test/extended/cli/builds.go
Comment (atiratree) [reply test/extended/cli/builds.go:77]: I wasn't sure with too many `It`s, since it generate a lot of new tests and `By` doesn't. I agree, it makes sense to use `It`s

File: test/extended/cli/builds.go
Comment (atiratree) [reply test/extended/cli/builds.go:59]: I am not sure if I should have AfterEach since I still haven't ported all the other tests and they will have different resources.

Or should  I just call `oc delete all --all` for simplicity there?

File: test/extended/cli/builds.go
Comment (atiratree) [reply test/extended/cli/builds.go:47]: we could add a check there

File: test/extended/cli/builds.go
Comment (atiratree) [reply test/extended/cli/builds.go:36]: Yes, I could add it. Please consider also https://github.com/openshift/origin/pull/26141#issuecomment-836731670

File: test/extended/cli/builds.go
Comment (atiratree) [reply test/extended/cli/builds.go]: removed the config and simplified the flow so `g.JustBeforeEach` is not needed anymore

File: test/extended/cli/builds.go
Comment (atiratree) [reply test/extended/cli/builds.go]: I have not researched the framework properly before. It seems it handles the deletion of the namespace in `AfterEach`, so additional logic should not be needed.

File: test/extended/cli/builds.go
Comment (atiratree) [reply test/extended/cli/builds.go]: I am using different functions now and I added checks for bcName empty, but it seems superfluous now in these...

File: test/extended/cli/builds.go
Comment (atiratree) [reply test/extended/cli/builds.go:151]: After a discussion changed to multiple standalone `It`s to support redability and better debugging.

**Comment by atiratree** [comment]
@ravisantoshgudimetla thanks for the review. I tried to fix all the problems, but still haven't ported all the logic yet.

**Comment by atiratree** [comment]
@ravisantoshgudimetla 

I have ported the first half of the `builds.sh`, but I am still struggling with waiting for the builds to complete. I cannot make any of the test builds successfully complete on my cluster. All end up in a Failed state. I will debug that more tomorrow.

Currently I am just waiting for the Build object to appear without checking the status/progress.

**Comment by ravisantoshgudimetla** [comment]
>  but I am still struggling with waiting for the builds to complete. I cannot make any of the test builds successfully complete on my cluster. All end up in a Failed state. I will debug that more tomorrow.

Not an expert myself but can you use [`StartBuildAndWait`](https://github.com/openshift/origin/blob/master/test/extended/builds/build_timing.go#L70) function and see if it works?

**Comment by soltysh** [comment]
> I found there is a bit of overlap with `[sig-builds]` as they are using oc cli to create builds as well. But it does not test the CLI part thourougly. I am not entirerly sure if it makes sense to combine `sig-builds` with `sig-cli`, but I decided not to at this moment.

@atiratree I forgot to mention, if you see duplicates feel free to remove as long as oc bits are properly tested in the existing tests. It's possible that for some of them it'll be true and it's ok to drop these tests, but some will be 1:1 duplicates, but the pre-existing will be using client-go, whereas yours will be strictly oc-focused - those should stay.

File: test/extended/cli/builds.go
Comment (soltysh) [reply test/extended/cli/builds.go]: There's no strict definition how to group them reasonable, I'll leave it up to you to decide so that they make a logical sense.

File: test/extended/cli/builds.go
Comment (soltysh) [inline test/extended/cli/builds.go:14]: Nit, but it's nice to split imports into the following logical groups:
1. built-in imports
2. 3rd party imports
3. openshift imports

You'll notice this being mostly applied to all of our source code and in k8s as well.

File: test/extended/cli/builds.go
Comment (soltysh) [reply test/extended/cli/builds.go:33]: First, you don't need to use `PollImmediate` since you're starting a build so I doubt it'll finish that soon, `wait.Poll` first waits and then performs poll which makes more sense in this case.

> I am not sure if I should wait here with `oc` or directly with go API

I'd probably go with an oc call, since we're focusing on that. The other tests are testing client-go calls.

File: test/extended/cli/builds.go
Comment (soltysh) [inline test/extended/cli/builds.go:48]: `%q` will be more appropriate, it'll handle quoting the string for you.

File: test/extended/cli/builds.go
Comment (soltysh) [inline test/extended/cli/builds.go:55]: Ditto.

File: test/extended/cli/builds.go
Comment (soltysh) [inline test/extended/cli/builds.go:35]: Unnecessary empty line, you're checking the error of the `Run` invocation, so it does make sense to group invocation and error check together. I see you're a fan of too many empty lines :wink:

File: test/extended/cli/builds.go
Comment (soltysh) [inline test/extended/cli/builds.go:67]: Drop emtpy line.

File: test/extended/cli/builds.go
Comment (soltysh) [inline test/extended/cli/builds.go:89]: Yeah, I agree with @ravisantoshgudimetla, I'd change those `It`'s to `By`'s.

Comment (soltysh) [reply test/extended/cli/builds.go:89]: Also, it's nice and simplifies review if the `By` description matches that from https://github.com/openshift/origin/blob/d704a4d2ab5e55731d11770c11eacd666940b944/test/extended/testdata/cmd/test/cmd/builds.sh#L31 which it does here, but not necessarily everywhere.

File: test/extended/cli/builds.go
Comment (soltysh) [inline test/extended/cli/builds.go:109]: I'm looking at the original test in https://github.com/openshift/origin/blob/d704a4d2ab5e55731d11770c11eacd666940b944/test/extended/testdata/cmd/test/cmd/builds.sh#L27-L29 and it doesn't wait for the build to succeed, actually, it only checks the definition of the newly created BuildConfig matches the expectation. Waiting for every build to complete unnecessarily lengthens the test time, which is not the goal. The build correctness is verified elsewhere, we care about oc here, mostly.

File: test/extended/cli/builds.go
Comment (soltysh) [reply test/extended/cli/builds.go:89]: My suggestion for organization is:
1. `Describe` as the top level
2. `It` as grouping, for example you have `new-build`, `complex-build`, etc.
3. `By` to mark every "group" from the original file

File: test/extended/testdata/application-template-dockerbuild.json
Comment (soltysh) [inline test/extended/testdata/application-template-dockerbuild.json:1]: Don't copy this file, it already lives here: https://github.com/openshift/origin/blob/master/examples/sample-app/application-template-dockerbuild.json re-use that instead. We currently have 2 copies already, my goal is to go to only a single one, b/c it's easier to maintain.

File: test/extended/testdata/cmd/test/cmd/builds.sh
Comment (soltysh) [inline test/extended/testdata/cmd/test/cmd/builds.sh:13]: When done, feel free to entirely drop this file.

**Review by soltysh — CHANGES_REQUESTED** [review]

File: test/extended/cli/builds.go
Comment (atiratree) [reply test/extended/cli/builds.go]: +1, looks much nicer

File: test/extended/cli/builds.go
Comment (atiratree) [reply test/extended/cli/builds.go]: will try to keep the empty lines to a reasonable amount :)  , hope it is better now

File: test/extended/cli/builds.go
Comment (atiratree) [reply test/extended/cli/builds.go]: I added one wait to the `start-build` test and removed the rest

File: test/extended/testdata/application-template-dockerbuild.json
Comment (atiratree) [reply test/extended/testdata/application-template-dockerbuild.json]: resolved

File: test/extended/cli/builds.go
Comment (soltysh) [inline test/extended/cli/builds.go:24]: We should be using images from here https://github.com/openshift/origin/blob/e87c32d71e0462e3c32ee96e4a198d087bc6f001/test/extended/util/image/image.go#L128 I think that `ShellImage()` should work here. The reason for that is that those images are then available for mirroring and thus for offline testing.

**Review by soltysh — APPROVED** [review]
/lgtm
/approve
That nit can be fixed in a followup.

**Comment by atiratree** [comment]
skpping the new tests on Disconnected clusters - the CI should pass now

/hold cancel

**Comment by soltysh** [comment]
/retitle Bug 1992596: e2e/cli: move half of builds.sh to proper e2e

**Review by soltysh — APPROVED** [review]
/lgtm

**Pull request merged**
